channeldb/invoices.go

```diff
@@ -1914,7 +1925,25 @@ func (d *DB) updateInvoice(hash *lntypes.Hash, invoices,
 	// the process by updating the state transitions for individual HTLCs
 	// and recalculate the total amount paid to the invoice.
 	var amtPaid lnwire.MilliSatoshi
-	for _, htlc := range invoice.Htlcs {
+	for key, htlc := range invoice.Htlcs {
+		// Set the HTLC preimage for any AMP HTLCs.
+		if setID != nil {
+			preimage, ok := update.State.HTLCPreimages[key]
+			switch {
+
+			// If we don't already have a preiamge for this HTLC, we
+			// can set it now.
+			case ok && htlc.AMP.Preimage == nil:
+				htlc.AMP.Preimage = &preimage
```

Comment:
it can be set in either the `Accepted` or `Settled` invoice level states. the following call to `updateHtlc` asserts that the correct preimage is present when attempting to transition to those states.